### DISCUSSION TITLE
json-prune doesn't work on uBO

### DISCUSSION THREAD

**Discussion by Yuki2718** [uBlockOrigin/uAssets#22556] [General] [answered]
https://www.msn.com/en-xl/news/other/21-famously-faulty-classic-cars/ss-AA19wOd5
`msn.com#%#//scriptlet('json-prune', 'configs.*.properties.slideshowWCSettings.interstitialNativeAds configs.*.properties.fullScreenSlideshowSettings.interstitialNativeAds properties.componentConfigs.slideshowConfigs.interstitialNativeAds')` works on AG, not on uBO. In the similar vein, `misskey.io##+js(json-prune, ads.[].imageUrl)` for https://misskey.io/@kureharu_stk works only on AG. I thought there was some changes made for yt which made incompatibility, but can these be fixed?

**Comment by Yuki2718** [comment]
For both cases `json-prune-fetch-response` works.

**Reply by gorhill** [reply]
What am I supposed to see on these pages without then with the filters? I don't see anything obvious without the filters, so I can't determine what the filters are supposed to do.

**Reply by Yuki2718** [reply]
Sorry, the problem were ads but as said solved by `json-prune-fetch-response` and I added the rules with it. So no need to spend your time. I should have written after testing more.

**Reply by gorhill** [reply]
But you said same filters work with AG but not uBO -- this is what I want to investigate.

For `https://misskey.io/@kureharu_stk`, stepping into the code I didn't see any instances of `ads.[].imageUrl`, the closest I saw was `ads.[].url`.

**Reply by Yuki2718** [reply]
It's logged only by `json-prune` The ads are `misskey.io##.xF6dY`.

![misskey](https://github.com/uBlockOrigin/uAssets/assets/58900598/780a3324-61d8-4006-8246-65da95957844)

**Reply by Yuki2718** [reply]
MSN ad is the same as https://github.com/AdguardTeam/AdguardFilters/issues/173054, blank slides. May need US IP. Found your comment I tried to mention in OP https://github.com/uBlockOrigin/uBlock-discussions/discussions/789#discussioncomment-6802445 (whoops, but does not open thx to be folded. Expand https://github.com/uBlockOrigin/uBlock-discussions/discussions/789#discussioncomment-6501001)

**Reply by gorhill** [reply, marked as answer]
The explanation is that uBO `json-prune` doesn't trap `Response.prototype.json` while [it does in AdGuard](https://github.com/AdguardTeam/Scriptlets/blob/master/src/scriptlets/json-prune.js#L137). uBO's `json-prune` used to also trap `Response.prototype.json`, but this was [changed in this commit](https://github.com/gorhill/uBlock/commit/749cec0f09) because a specialized `json-prune-fetch-response` was spun off to avoid trapping unnecessarily. AdGuard doesn't have `json-prune-fetch-response`.

**Reply by Yuki2718** [reply]
Yeah, that's what I forgot and initially confused. But shouldn't `json-prune-fetch-response` log `ads.[].imageUrl` and `json-prune` not?

**Reply by gorhill** [reply]
I didn't see `imageUrl` reported by `+js(json-prune)`. As for `+js(json-prune-fetch-response)`, looking at the code, it's not set up to log like `json-prune` -- I will try to remember to support this in current dev cycle.

**Reply by gorhill** [reply]
https://github.com/gorhill/uBlock/commit/e527a8f9af7f9ca8f25f0cff19646d77d85a4edb

**Comment by Yuki2718** [comment]
`www.msn.com##+js(json-prune-fetch-response, properties.componentConfigs.slideshowConfigs.slideshowSettings.interstitialNativeAds, , propsToMatch, url:consumptionpage/gallery_windows/config.json)` removes ad placeholder in slides on `https://www.msn.com/en-us/health/wellness/18-american-phrases-that-non-americans-struggle-to-understand/ss-AA1nbhpn?ocid=winp1taskbar&cvid=ad2460971cdc406fa2002b8e2dd27977&ei=6#image=21` but nothing about `interstitialNativeAds` is logged with 
```
msn.com##+js(json-prune)
msn.com##+js(json-prune-fetch-response)
```
Same with v1.54.

**Reply by gorhill** [reply]
Sorry, not understanding what you are reporting.

**Reply by Yuki2718** [reply]
Just wondering why `interstitialNativeAds` never appers in logs while pruning this fixes ad placeholder.

**Reply by gorhill** [reply]
Is the pruning really being applied? I don't see it being reported in the logger, and placeholders are not removed.

>     msn.com##+js(json-prune-fetch-response)
> why `interstitialNativeAds` never appers in logs

I do see one reported line:

![image](https://github.com/uBlockOrigin/uAssets/assets/585534/0af039e6-bacb-4844-911c-a4d5c71f8add)

---

Note that on my side the JSON paths to `interstitialNativeAds` are:

    properties.slideshowWCSettings.interstitialNativeAds
    properties.fullScreenSlideshowSettings.interstitialNativeAds

**Reply by Yuki2718** [reply]
They are old paths (see https://github.com/AdguardTeam/AdguardFilters/commit/f2dd575567601277f152966679f225867a3731e7) and on my end no more used.
